### PR TITLE
support class center sample of PartialFC

### PR DESCRIPTION
**Pull request by GuoxiaWang**
### PR types
New features

### PR changes
APIs

### Describe

class center sample of PartialFC

* Support CPU
* Support single GPU
* Support multi GPU

Class center sample method is proposed from the paper PartialFC that only sample a subset of the class centers. The process of sampling subset class centers is straightforward: 1) First select the positive class centers; 2) Randomly sample negative class centers.

For more information, Partial FC: Training 10 Million Identities on a Single Machine. arxiv: https://arxiv.org/abs/2010.05222

CPU:
For CPU, we can use std::set and construct a lookup table very easy. So we first get unique positive class center by ascending. If num_positive_class_center < num_samples, then sample negative class center randomly. Finally, we can get the remapped label by remap the input label according sampled class center.

GPU:
We first randomly generate a value in [0, num_classes) on each position in a array(shape[num_classes]). Then, we mark the element as negative value in the array according input label. Now, we can sort the array by ascending to ensure that the positive class center always in the front of the sorted array. So, we can get the sampled class center index by sorted keys. Finally, we can get the remapped label by remap the input label according sampled class center.

### CODE REVIEW

File: paddle/fluid/operators/class_center_sample_op.cc [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+)DOC");
+  }
+};
+
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+namespace plat = paddle::platform;
+REGISTER_OP_WITHOUT_GRADIENT(class_center_sample, ops::ClassCenterSampleOp,
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cc:143]: 该OP为什么没有反向？

Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cc:143]: 该OP为什么没有反向

File: paddle/fluid/operators/class_center_sample_op.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    int seed = ctx.Attr<int>("seed");
+    bool fix_seed = ctx.Attr<bool>("fix_seed");
+    PADDLE_ENFORCE_GT(num_classes, 0,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_classes' for Op(class_center_sample) "
+                          "must be greater than 0, "
+                          "but the value given is %d.",
+                          num_classes));
+
+    PADDLE_ENFORCE_GT(num_sample, 0,
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.h]: 使用num_samples更符合规范

File: paddle/fluid/operators/class_center_sample_op.cu [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    }
+    index[i] = static_cast<T>(i);
+  }
+}
+
+template <typename T>
+__device__ void FindIntervalIndex(const T* class_interval_ptr,
+                                  const int64_t nranks, const int value,
+                                  int64_t* find_index) {
+  int64_t sta = 0;
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cu]: may be start is much easier to understand

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -14839,6 +14840,150 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
     return out
 
 
+def class_center_sample(label, num_classes, num_sample, group=None, seed=None):
+    """
+    Class center sample method is proposed from the paper PartialFC that only sample a subset of the class centers.
+    The process of sampling subset class centers is straightforward: 1) First select the positive class centers;
+    2) Randomly sample negative class centers. Specifically, given a label tensor, shape [batch_size], select all
+    the positive class centers and randomly sample negative class centers, then remap the input label tensor using
+    the sampled class centers. Note that if the number of the positive class centers greater than the input
```
Comment (sandyhouse) [inline python/paddle/fluid/layers/nn.py]: is greater than

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -14839,6 +14840,150 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
...
 
 
+def class_center_sample(label, num_classes, num_sample, group=None, seed=None):
+    """
+    Class center sample method is proposed from the paper PartialFC that only sample a subset of the class centers.
+    The process of sampling subset class centers is straightforward: 1) First select the positive class centers;
+    2) Randomly sample negative class centers. Specifically, given a label tensor, shape [batch_size], select all
+    the positive class centers and randomly sample negative class centers, then remap the input label tensor using
+    the sampled class centers. Note that if the number of the positive class centers greater than the input 
+    num_sample, it keep all the positive class centers and the shape of sampled_class_center will be
```
Comment (sandyhouse) [inline python/paddle/fluid/layers/nn.py]: it keeps

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -14839,6 +14840,150 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
...
+    the positive class centers and randomly sample negative class centers, then remap the input label tensor using
+    the sampled class centers. Note that if the number of the positive class centers greater than the input 
+    num_sample, it keep all the positive class centers and the shape of sampled_class_center will be 
+    [num_positive_class_centers]. The API supports CPU, single GPU and multi GPU.
+
+    For more information, Partial FC: Training 10 Million Identities on a Single Machine
+    arxiv: https://arxiv.org/abs/2010.05222
+
+    Args:
+    	label: label list, each label in [0, num_classes)
```
Comment (sandyhouse) [inline python/paddle/fluid/layers/nn.py]: 规范格式，比如 label (list):

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -14839,6 +14840,150 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
...
+    if core.is_compiled_with_dist():
+        parallel_env = paddle.distributed.ParallelEnv()
+        global_rank = parallel_env.rank
+        rank = global_rank if group is None else group.get_group_rank(
+            global_rank)
+        nranks = parallel_env.world_size if group is None else group.nranks
+
+    if num_sample > num_classes:
+        raise ValueError(
+            'Expected num_sample less equal than {}, got num_sample {}'.format(
```
Comment (sandyhouse) [inline python/paddle/fluid/layers/nn.py]: be less than or equal to, or just <=

File: python/paddle/fluid/tests/unittests/parallel_class_center_sample.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    return remapped_label, pos_class_center_per_device
+
+
+class TestParallelClassCenterSampleOp(unittest.TestCase):
+    def setUp(self):
+        strategy = fleet.DistributedStrategy()
+        self.model_parallel_size = 2
+        strategy.hybrid_configs = {
+            "dp_degree": 1,
+            "mp_degree": self.model_parallel_size,
```
Comment (sandyhouse) [inline python/paddle/fluid/tests/unittests/parallel_class_center_sample.py]: why use model parallel?

File: paddle/fluid/operators/class_center_sample_op.cu [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+    GetSampledClassCenter<T><<<NumBlocks(batch_size), kNumCUDAThreads, 0,
+                               ctx.cuda_device_context().stream()>>>(
+        actual_num_sample, cub_sort_values_out_ptr,
+        sampled_local_class_center_ptr);
+  }
+};
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cu:481]: 为什么没有反向实现？

File: paddle/fluid/operators/class_center_sample_op.cu [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+    PADDLE_ENFORCE_GT(num_classes, 0,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_classes' for Op(class_center_sample) "
+                          "must be greater than 0, "
+                          "but the value given is %d.",
+                          num_classes));
+
+    PADDLE_ENFORCE_GT(num_sample, 0,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_sample' for Op(class_center_sample) "
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cu]: num_samples?

File: paddle/fluid/operators/class_center_sample_op.cu [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+                          "must be greater than 0, "
+                          "but the value given is %d.",
+                          num_sample));
+
+    PADDLE_ENFORCE_LE(num_sample, num_classes,
+                      platform::errors::InvalidArgument(
+                          "The value 'num_sample' for Op(class_center_sample) "
+                          "must be less than %d, "
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cu]: <=? or use PADDLE_ENFORCE_LT

File: paddle/fluid/operators/class_center_sample_op.cu [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+          static_cast<platform::CUDADeviceContext*>(
+              platform::DeviceContextPool::Instance().Get(ctx.GetPlace()))
+              ->stream();
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclAllReduce(
+          num_classes_per_device_ptr, num_classes_per_device_ptr,
+          num_classes_per_device.numel(),
+          platform::ToNCCLDataType(num_classes_per_device.type()), ncclSum,
+          comm->comm(), calcu_stream));
+    }
+#endif
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cu:344]: What happens when PADDLE_WITH_NCCL or PADDLE_WITH_RCCL is not defined?

File: paddle/fluid/operators/class_center_sample_op.cu [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+
+    // step 2: Determine temporary device storage requirements
```
Comment (sandyhouse) [inline paddle/fluid/operators/class_center_sample_op.cu:346]: You'd better to give an illustration on how the algorithm works in the pr description.

File: python/paddle/fluid/layers/nn.py [hunk 14]
```diff
@@ -14839,6 +14840,150 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
...
+    the sampled class centers. Note that if the number of the positive class centers greater than the input 
+    num_sample, it keep all the positive class centers and the shape of sampled_class_center will be 
+    [num_positive_class_centers]. The API supports CPU, single GPU and multi GPU.
+
+    For more information, Partial FC: Training 10 Million Identities on a Single Machine
+    arxiv: https://arxiv.org/abs/2010.05222
+
+    Args:
+    	label: label list, each label in [0, num_classes)
+    	num_classes: A positive integer to specify the number of classes at local rank.
```
Comment (sandyhouse) [inline python/paddle/fluid/layers/nn.py]: num_classes指得是每个卡自己的类别数？这样是不是和语义有些冲突。更合理的建议是设置全局类别数，通过内部计算得到每张卡的类别数。

**Review by sandyhouse — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/class_center_sample_op.cc (hunk 1 above)
Comment (GuoxiaWang) [reply paddle/fluid/operators/class_center_sample_op.cc:143]: 这个Op不需要反向，它的作用只是采样出类别中心，有了类别中心索引，结合 gather 就可以把相应的列取出来，参数在反向网络更新时 gather 会有反向更新参数的操作。

File: paddle/fluid/operators/class_center_sample_op.h (hunk 2 above)
Comment (GuoxiaWang) [reply paddle/fluid/operators/class_center_sample_op.h]: > 使用num_samples更符合规范

好的

File: paddle/fluid/operators/class_center_sample_op.cu (hunk 11 above)
Comment (GuoxiaWang) [reply paddle/fluid/operators/class_center_sample_op.cu]: 是 <=， 输出忘记没改过来。这里最开始设定是 < 的，但是为了后续测试，<= 可以等价为 sample_ratio = 1.0

File: paddle/fluid/operators/class_center_sample_op.cu (hunk 12 above)
Comment (GuoxiaWang) [reply paddle/fluid/operators/class_center_sample_op.cu:344]: 没有的话，就是单机单卡，既然都没有编译分布式，那么必然不能初始化 fleet，所以不存在多卡通信

File: python/paddle/fluid/layers/nn.py (hunk 14 above)
Comment (GuoxiaWang) [reply python/paddle/fluid/layers/nn.py]: 局部 rank 的类别数更合理，局部的，在算子里通信就可以得到全局的，同时局部类别数可以不一致，天然支持不均等切分。如果是全局的，想在内部算出每张卡被别数，基本可以认为需要均分切分类别数，最多最后一张卡类别数可以和其他卡不一样。

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by TCChenlong — DISMISSED** [review]

**Review by sandyhouse — DISMISSED** [review]
LGTM

File: python/paddle/fluid/layers/nn.py [hunk 15]
```diff
@@ -14840,6 +14841,164 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
     return out
 
 
+def class_center_sample(label, num_classes, num_samples, group=None, seed=None):
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/layers/nn.py]: 这些代码需要放到python/paddle/nn/functionl目录下，
fluid在paddle2.0不推荐用户使用，原则上不在python/paddle/fluid目录增加新的代码

File: python/paddle/fluid/layers/nn.py [hunk 16]
```diff
@@ -14840,6 +14841,164 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
...
+    .. code-block:: python
+
+        # CPU or single GPU
+        import os
+        import paddle
+        import numpy as np
+        num_classes = 20
+        batch_size = 10
+        num_samples = 6
+        np_label = np.random.randint(0, num_classes, (batch_size,), dtype=np.int64)
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/layers/nn.py]: 这里推荐使用paddle.randint，避免引入numpy

File: python/paddle/fluid/layers/nn.py [hunk 17]
```diff
@@ -14840,6 +14841,164 @@ def shard_index(input, index_num, nshards, shard_id, ignore_value=-1):
...
+        import paddle.distributed as dist
+        import numpy as np
+        strategy = dist.fleet.DistributedStrategy()
+        dist.fleet.init(is_collective=True, strategy=strategy)
+        batch_size = 10
+        num_samples = 6
+        rank_id = dist.get_rank()
+        # num_classes of each GPU can be different, e.g num_classes_list = [10, 8]
+        num_classes_list = [10, 10]
+        num_classes = np.sum(num_classes_list)
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/layers/nn.py]: 优先使用paddle的原生API paddle.sum或 paddle.randint

**Review by XiaoguangHu01 — CHANGES_REQUESTED** [review]

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by TCChenlong — DISMISSED** [review]

**Review by XiaoguangHu01 — DISMISSED** [review]
LG API

**Review by raindrops2sea — DISMISSED** [review]

**Review by XieYunshen — DISMISSED** [review]
LGTM

**Review by cryoco — DISMISSED** [review]

**Review by gongweibao — DISMISSED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]

**Review by cryoco — APPROVED** [review]

**Review by sandyhouse — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**